### DISCUSSION TITLE
Early directory tree pruning with hive-partitioned `parquet_scan`

### DISCUSSION THREAD

**Discussion by mvanaltvorst** [duckdb/duckdb#7620] [Ideas]
I occasionally work with 1TB+ hive-partitioned parquet datasets where I want to extract small 10GB chunks to process. An issue I encounter with duckdb is that despite of the `WHERE` constraints I use, I seem to suffer from a 30 second fixed time overhead.

I believe @samansmink has described the main cause and a possible solution here: https://github.com/duckdb/duckdb/pull/4746#issuecomment-1250925570. By not fetching every filename in advance, but recursively applying filters as you encounter them in the file system structure, you can make sure you make optimal use of the hive-partitioned data structure.

I've experimented with this before with a hacky Python script, and it is definitely a lot quicker than anything `duckdb` or `pyarrow.parquet.ParquetDataset` have to offer. If duckdb would support this, it would be the only framework I know that can handle 1TB+ hive-partitioned datasets with ease.

**Comment by mvanaltvorst** [comment]


**Reply by Tishj** [reply]
I think lverdoes mentioned this recently.
What you're talking about is essentially pushing down the filter into the glob right? I believe they are going to start working on this soon if that's the case.

If it's not then disregard this comment

**Comment by samansmink** [comment]
@lverdoes is indeed working on this one!

**Comment by outinspace** [comment]
I'm curious if there's any update on this optimization. I'm interested in this optimization so that I can use DuckDb as a multi-tenant reporting engine, where single-tenant queries are not affected by the number of files in other tenant partitions.

Related: https://github.com/duckdb/duckdb/discussions/13614

**Comment by samansmink** [comment]
@outinspace Latest update is the fact that we have added multifilelists which allow pushing down filters into filelists in a clean way: https://github.com/duckdb/duckdb/pull/11806

Basically we are slowly moving in a way where we can cleanly implement this, but i can't give a concrete timeline for this feature yet

**Comment by xevix** [comment]
@samansmink I would be interested in giving this feature a try, unless it conflicts with the existing DuckDB roadmap, or it's better if a core developer works on it.

For now I've worked around it by adding the Hive key to the path and out of the WHERE clause, which works well but is less ergonomic. I hit this when running data on a local disk rather than S3, and even there it adds 300ms if using `**/*.parquet`, 200ms when using `*/*/*.parquet` which covers the two Hive keys in use in this dataset. Removing them all and reading one top-level partition is just 20ms.

**Reply by samansmink** [reply]
Hey @xevix! I think you should be able to pick this one up. Basically what you want to do is either to expand the existing `GlobMultiFileList` or create a new `HivePartitionedMultiFileList` class that allow passing in an option for pruning the directory tree using the filters on hive partitions. So using the filters that are passed [here](https://github.com/duckdb/duckdb/blob/e0a2d33eafabb9c7b61d88ac334ac7c685bfc4e4/src/include/duckdb/common/multi_file/multi_file_list.hpp#L162) you should be able to more effectively glob the files [here](https://github.com/duckdb/duckdb/blob/e0a2d33eafabb9c7b61d88ac334ac7c685bfc4e4/src/common/multi_file/multi_file_list.cpp#L372)

Let me know if you have any questions!

**Reply by xevix** [reply]
Awesome, I'll give it a go!

I'm thinking that the FileList would attempt to apply hive key filters, and if they succeed, only list files in the matching directories, and recursively dive down the hierarchy. One thing [here](https://github.com/duckdb/duckdb/blob/977e76d3bffc227d4d31f9ce0c300e01a4b84003/src/common/multi_file/multi_file_reader.cpp#L239) is that the verification of partitions iterates through all files, but I think it should be sufficient to check the structure _only_ for directories whose contents will actually be listed, which should avoid the extra list operations. Perhaps there could be an option to force enabling/disabling of Hive validation. I think if a Hive key is in the filters, we could disable this behavior by default, but in the absence of this, check anyway. In addition, we would check at all levels of the directory tree to which filters have not applied.

Let me know if the following sounds ok

## Example

So for example with 3 keys in the structure: `data/YEAR=2025/MONTH=7/DAY=1`, a query could look like:

```sql
FROM 'data/**/*.parquet' WHERE year = 2025 AND day = 1
```

**Current directory: `data/`**

It would start at the `data` dir, successfully apply the year filter, and as a result, not list all the folders for all the years for Hive validation, but simply move on.

**Current directory: `data/YEAR=2025/`**

No Hive key filter can be applied here since we're not filtering by month, so we are forced to list all available months, and we validate the Hive structure in the process, then go a level deeper.

**Current directories: `data/YEAR=2025/MONTH=*`**

For each month directory, we try to filter by Hive keys, and successfully filter by day in each case, and as a result, we don't list or validate Hive for other day folders.

**Current directories: `data/YEAR=2025/MONTH=*/DAY=1`**

Finally, we reach the folders with the parquet files.

**Reply by xevix** [reply]
I think to make this work, we'd have to make path expansion lazy, which seems like it was already planned, so I hope I don't duplicate anybody's WIP here. This way, we can step breadth-first down the filesystem tree one step at a time, prune the file list, then recursively expand as usual.

https://github.com/duckdb/duckdb/blob/977e76d3bffc227d4d31f9ce0c300e01a4b84003/src/common/multi_file/multi_file_list.cpp#L277-L287

**Reply by samansmink** [reply]
Don't worry about duplicating anyones work, AFAIK nobody has taken a stab at this yet.

You can likely just delay the glob expansion till the point where the first path is requested. Note that the pruning step using `HivePartitioning::ApplyFiltersToFileList`  will still need to be done, but this currently takes a fully materialized list only. You can either look into rewriting that to allow applying it per file, or fully materialize the whole list once the first file is requested. I think the latter will work, but will require some testing to be sure. This might need some other changes to the `GlobMultiFileList`, like the `GlobMultiFileList::GetExpandResult` which currently expands the first file. I think this can just be changed to always return `MULTIPLE_FILES`, like delta already [does](https://github.com/duckdb/duckdb-delta/blob/61813660c182fa95367af35b39228fb16ba1ec48/src/functions/delta_scan/delta_multi_file_list.cpp#L958).

**Reply by xevix** [reply]
So currently I believe the `Glob()` method, if given a `**` will recurse exhaustively, so if we wanted to delay glob expansion, would either have to rewrite `Glob()` for all filesystems, or change the way that `Glob()` is called from the caller.

Example query:

```
FROM 'data/**/*.parquet' WHERE year IN (2024, 2025) AND day = 1
```

If changing calls to `Glob()`:

We could use the filters to reconstruct the paths and attempt: `Glob("data/YEAR=2024")`, `Glob("data/YEAR=2025")`, `Glob("data/DAY=1")`. If any of these globs succeeds, we call Glob even deeper, e.g. `Glob("data/YEAR=2025/DAY=1")`. If this fails at any level for all possible filters, we add * and go one level down, e.g. `Glob("data/YEAR=2025/*/DAY=1")`. Then if all filters are exhausted, we just add the remaining suffix of the path, e.g. `Glob("data/YEAR=2025/*/DAY=1/*.parquet")`. This is tricky, but avoids having to rewrite all the `Glob()` implementations to allow getting paths 1 level of depth at a time. I'm hoping to avoid materializing the full (unfiltered) list of files for this optimization, but this would require being able to filter somehow one level at a time of the filesystem tree.

Let me know if I misunderstood what you're suggesting or if I'm missing something critical in my thought process here. In any case, this does seem like a fairly involved rewrite, so I want to make sure of the approach before diving head-first in. Thank you for helping to think through this and give design feedback, I majorly appreciate it.

**Reply by samansmink** [reply]
Well since the glob expansion is filesystem specific, different filesystems could have different glob semantics or syntax. I would say we need to push down the hive partition filters into the FileSystem::GlobFiles() call, then each filesystem would need to be modified to use these filters

**Reply by xevix** [reply]
That's an interesting point. In the [glob docs](https://duckdb.org/docs/stable/data/multiple_files/overview#glob-syntax) it shows the glob syntax as filesystem agnostic. I assumed that given the generic handling of paths by the existing Hive filters that there's some assumption that all filesystems are hierarchical, use separators between each directory, and so share the same semantics of recursively traversing them. I was a bit reluctant to move the Hive filters down into `FileSystem::Glob()` because of separation of concerns (`FileSystem` getting files, vs. `MultiFileList` applying filters to the files), but maybe it's easier to have filesystems run the filters as part of globbing, rather than changing `FileSystem::Glob()` to return directories one depth level at a time.

I'm going to give this a bit more thought but I have enough to start now. Thanks!

Update: Ah, seems like [S3 globbing](https://duckdb.org/docs/stable/core_extensions/httpfs/s3api.html#globbing) only supports `*` and not `**`. I can implement pushing the filters into `glob` for core filesystems, but this will break 3rd party filesystem extensions I think. I could maybe add a boolean to filesystem to indicate if a filesystem supports pushdown Hive pruning, and have a different Glob function that gets called in that case that can handle it, to avoid breaking everything.

**Reply by samansmink** [reply]
I think you'd probably get away with implementing this on the GlobMultiFileList side, since the glob patterns are so similar across filesystems, but I think it's cleaner to let each filesystem handle this individually.

Perhaps we can add a `GlobFiltered(const string &path, GlobFilters filters, FileOpener *opener = nullptr)` which has a default implementation of just calling the regular `Glob` function ignoring the filters?

**Reply by xevix** [reply]
Another consideration is there's a few places such as in [binding](https://github.com/duckdb/duckdb/blob/c2349afa7585cbc2e9560176ae4fc7810ae662df/src/common/multi_file/multi_file_reader.cpp#L125) or [Hive detection](https://github.com/duckdb/duckdb/blob/c2349afa7585cbc2e9560176ae4fc7810ae662df/src/common/multi_file/multi_file_reader.cpp#L747) that end up triggering the eager globbing. Since this is before the optimizer's filter pushdown, we can't apply Hive filters at that point. One thing could be to make globbing lazy, or to special case this common use case of needing to know if there's at least 1 file present, and having 1 file path just to auto-detect Hive partitioning and so forth, and just provide one file in the globber and return immediately if given a flag or so. Then when the optimizer pushdown happens later, at that point we can restart the expansion of the files, ignoring the expanded file from before, but this time we have Hive filters. Does this sound reasonable?

EDIT: Sorry, it seems like you already addressed this earlier when you referenced Delta. It's been taking me a bit to go over all the affected code and changes, thanks for your patience.

Separately, one advantage of implementing the actual filtering per-filesystem is that for local filesystem, we could attempt filtering at each level which is a cheap operation (e.g. listing `YEAR=2024/` if it's a hive filter), but for S3 where listing files has a significant round-trip cost we could avoid trying the filters at each stage.

**Reply by samansmink** [reply]
@xevix no worries, this is definitely not an easy feature to implement :)

Yea indeed in delta we already achieve the lazy filelist expansion, so it should be possible here too. Let me know if you have any more questions, happy to help!

**Reply by xevix** [reply]
Thanks, I appreciate all the help. I have a better understanding now, and am trying to think through the best approach that doesn't break too much encapsulation. Right now, things like Hive validation are happening in the binder, which forces path expansion, and feels like it's executing things by doing the actual path lists involving I/O calls. How much execution is permissible in the binder, as opposed to in physical operators?

I'm going to start by just optimizing for Hive-partitioned parquet files in local and S3/httpfs filesystems. Functions like `GetTotalFileCount()` are called when explicit_cardinality is set (I think for Iceberg?), and in JSON, CSV etc., but maybe that can be addressed in follow-up work.

Rather than going full lazy evaluation in glob with iterator returning one file at a time, I'm thinking to have a `max_files` param which stops after a certain number of files, to match `MultiFileList::GetFirstFile()` and `MultiFileList::GetFile(idx_t i)` which just need to peek at the first files without needing the full set, for things like column binding. This reduces the code changes needed, including the way function callers handle the results. I'm on the fence for whether Hive filters should be passed down the call stack to reach `Glob()` or whether to add them into the `MultiFileList` as options that can be set before calling `Glob()`. Passing down the call stack is a bit messy and requires changing all filesystems' `Glob` signature, but might be clearer.

**Reply by xevix** [reply]
Is there a way to do later binding of columns and types, to allow hive filtering in the optimizer? Originally was thinking to stop at first file found for columns and types, but `union_by_name` would not work with this unless we can delay binding until after hive filtering.

**Reply by samansmink** [reply]
> union_by_name would not work with this

I think having `union_by_name` not work with this is fine

> Functions like GetTotalFileCount() are called when explicit_cardinality is set

Yea part of this feature would be to ensure that we don't unnecessarily materialize the file list. I could see how this feature would require a specific configuration to be fully effective, e.g. by requiring that explicit_cardinality is set to false. This could be either done through documentation and/or an `InvalidConfigurationException` when the settings are incompatible

> Rather than going full lazy evaluation in glob with iterator returning one file at a time, I'm thinking to have a max_files param

Yea I think this makes sense. So something like this:
- Create a GlobMultiFileList which has an option to start with a "cheap" glob that lists max n files.
- The first n files are fetched from the GlobMultiFileList and used during binding to check the hive schema
- Pushdown is performed and creates a new multifilelist
- Query execution starts and begins lazily expanding the multifilelist as it is scanned

**Reply by xevix** [reply]
I've gotten a basic example working at a very basic level, but the code is nowhere near done or ready for review, still very WIP. The number of refactors was a lot heavier than I had originally planned so I might break this down into a few pieces:

* `LocalFileSystem::ListFilesExtended()`: support early stop of listing files signaled by a boolean from the callback (avoids unnecessary file scanning)
* `GetFirstFile()` et al: support early returns once `max_files` (i.e. 1 file in this case) have been found. Use this for Hive partitions checks as well as initial Hive type checks
* Support pushing down Hive filters into `Glob()`
* Postpone checking Hive consistency until after Hive filters applied
* Fall back to eager file listing if conditions are not met (e.g. `union_by_name` or `explicit_cardinality` set)

DuckDB 1.3.2
```
EXPLAIN ANALYZE SELECT DISTINCT ID FROM '/Users/xevix/Downloads/data/noaa/by_year/**/*.parquet' WHERE YEAR = 2024 AND ELEMENT = 'TMAX' ORDER BY ALL;
┌────────────────────────────────────────────────┐
│┌──────────────────────────────────────────────┐│
││              Total Time: 0.419s              ││
│└──────────────────────────────────────────────┘│
└────────────────────────────────────────────────┘
```

Hive filter pushdown in glob
```
EXPLAIN ANALYZE SELECT DISTINCT ID FROM '/Users/xevix/Downloads/data/noaa/by_year/**/*.parquet' WHERE YEAR = 2024 AND ELEMENT = 'TMAX' ORDER BY ALL;
┌────────────────────────────────────────────────┐
│┌──────────────────────────────────────────────┐│
││              Total Time: 0.0153s             ││
│└──────────────────────────────────────────────┘│
└────────────────────────────────────────────────┘
```

I'll keep working at this but it's likely going to take a bit longer to come up with something cleaner and fully correct. Thanks for the patience.

WIP: https://github.com/duckdb/duckdb/compare/main...xevix:duckdb:hive-directory-pruning-2025-07-14?expand=1

**Reply by xevix** [reply]
I tried a few different approaches like going full lazy with Iterator classes just in case, but I don't think the performance difference is worth the complexity. Current approach is a refactor of the Hive filters to work on a single file, as well as a full list. This way we can consistently support the case where optimization isn't possible (e.g. union_by_name), as well as when optimization is possible. I made `HivePartitioning` into an instance which will allow keeping track of overall state even as we call the single-file method one file at a time from multiple parts of the Glob methods.

One consequence is that total files in `extra_info` will be less than helpful, since we cannot know the total number of files, having filtered out directories before having a chance to count them, but I think this is an acceptable loss of information, and we can always force this optimization off if people really want to get the stats.

Next up is detecting and supporting when optimization isn't possible and keeping that case. Hopefully this is an ok approach.

https://github.com/xevix/duckdb/blob/hive-directory-pruning-2025-07-14/src/include/duckdb/common/hive_partitioning.hpp

```
D SELECT DISTINCT ID FROM '/Users/xevix/Downloads/data/noaa/by_year/**/*.parquet' WHERE YEAR = 2024 AND ELEMENT = 'TMAX' ORDER BY ALL;
┌─────────────────────────────────────┐
│┌───────────────────────────────────┐│
││    Query Profiling Information    ││
│└───────────────────────────────────┘│
└─────────────────────────────────────┘
SELECT DISTINCT ID FROM '/Users/xevix/Downloads/data/noaa/by_year/**/*.parquet' WHERE YEAR = 2024 AND ELEMENT = 'TMAX' ORDER BY ALL;
┌────────────────────────────────────────────────┐
│┌──────────────────────────────────────────────┐│
││              Total Time: 0.0154s             ││
│└──────────────────────────────────────────────┘│
└────────────────────────────────────────────────┘
┌────────────────────────────────────────────────┐
│               Optimizer: 0.0058s               │
│┌──────────────────────────────────────────────┐│
││        Build Side Probe Side: 0.0000s        ││
││           Column Lifetime: 0.0000s           ││
││           Common Aggregate: 0.0000s          ││
││        Common Subexpressions: 0.0000s        ││
││      Compressed Materialization: 0.0000s     ││
││          Cte Filter Pusher: 0.0000s          ││
││             Cte Inlining: 0.0000s            ││
││             Deliminator: 0.0000s             ││
││           Duplicate Groups: 0.0000s          ││
││         Empty Result Pullup: 0.0000s         ││
││         Expression Rewriter: 0.0000s         ││
││              Extension: 0.0000s              ││
││            Filter Pullup: 0.0000s            ││
││           Filter Pushdown: 0.0057s           ││
││              In Clause: 0.0000s              ││
││         Join Filter Pushdown: 0.0000s        ││
││              Join Order: 0.0000s             ││
││         Late Materialization: 0.0000s        ││
││            Limit Pushdown: 0.0000s           ││
││           Materialized Cte: 0.0000s          ││
││             Regex Range: 0.0000s             ││
││            Reorder Filter: 0.0000s           ││
││          Sampling Pushdown: 0.0000s          ││
││        Statistics Propagation: 0.0000s       ││
││             Sum Rewriter: 0.0000s            ││
││                Top N: 0.0000s                ││
││           Unnest Rewriter: 0.0000s           ││
││            Unused Columns: 0.0000s           ││
│└──────────────────────────────────────────────┘│
└────────────────────────────────────────────────┘
┌────────────────────────────────────────────────┐
│            Physical planner: 0.0000s           │
│┌──────────────────────────────────────────────┐│
││            Column Binding: 0.0000s           ││
││             Create Plan: 0.0000s             ││
││            Resolve Types: 0.0000s            ││
│└──────────────────────────────────────────────┘│
└────────────────────────────────────────────────┘
┌────────────────────────────────────────────────┐
│                Planner: 0.0026s                │
│┌──────────────────────────────────────────────┐│
││               Binding: 0.0026s               ││
│└──────────────────────────────────────────────┘│
└────────────────────────────────────────────────┘
┌───────────────────────────┐
│           QUERY           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│          ORDER_BY         │
│    ────────────────────   │
│           ID ASC          │
│                           │
│         12920 Rows        │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│       HASH_GROUP_BY       │
│    ────────────────────   │
│         Groups: #0        │
│                           │
│         12920 Rows        │
│          (0.01s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         PROJECTION        │
│    ────────────────────   │
│             ID            │
│                           │
│        3633379 Rows       │
│          (0.00s)          │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│         TABLE_SCAN        │
│    ────────────────────   │
│         Function:         │
│        PARQUET_SCAN       │
│                           │
│      Projections: ID      │
│                           │
│       File Filters:       │
│  (YEAR = 2024)(ELEMENT =  │
│          'TMAX')          │
│                           │
│    Scanning Files: 8/8    │
│    Total Files Read: 8    │
│                           │
│        3633379 Rows       │
│          (0.02s)          │
└───────────────────────────┘
```

**Reply by samansmink** [reply]
> One consequence is that total files in extra_info will be less than helpful, since we cannot know the total number of files, having filtered out directories before having a chance to count them, but I think this is an acceptable loss of information, and we can always force this optimization off if people really want to get the stats.

Yea I think this is a fundamental problem right now and something to leave to a future PR. Maybe in cases where we know we don't know the full file list size, we could omit the `Scanning Files: 8/8` files since it's confusing, and only print them when we do know for sure the total count.

Note that in delta I added an option specifically for this: https://github.com/duckdb/duckdb-delta/blob/62782140e00e095bf295e92a72e2ee1192502cc7/src/delta_extension.cpp#L74

**Reply by xevix** [reply]
Good call, I should remove the Scanning Files in this case.

To try to avoid rewriting the `ListFilesExtended` method to be lazy which might be a heavy lift, I'm thinking of changing `Glob` so that instead of a for loop for each split, we recursively call a `ProcessSplit()` method that looks similar to the body of the for loop, but will allow `RecursiveGlobDirectories()` to call `ProcessSplit()` once for each directory it encounters immediately. This will allow us to get to files immediately, and allow us to check for `max_files` so we can break early once we have enough files. Let me know if there's concerns with the stack with this much recursion, or if you think it makes sense to pursue a deeper rewrite of `ListFilesExtended` to be fully lazy.

I think laziness doesn't help a ton, because there's essentially only a few functions that need to return early
* GetFirstFile()
* GetFile(1) // grabs first 2 files to see if there's many files

If I implement the above recursion, then we should be able to keep the `GetExpandResult()` rather than just returning ManyFiles as in Delta, and it would be less hacky than a specific internal check for just the 2 cases of "get first file" and "get first 2 files."

**Reply by xevix** [reply]
I rewrote the Glob() to use mutual recursion. The performance of the lazy Hive filters is still great at 20ms. The performance with `union_by_name=true` went down from 700ms+ to ~600ms.

Code: https://github.com/duckdb/duckdb/compare/main...xevix:hive-directory-pruning-2025-07-23?expand=1

**Reply by xevix** [reply]
Working on fixing tests and addressing bugs and edge cases, down to about 5 tests. For example `CSVSchemaDiscovery` calls `GetAllFiles()` which loads files eagerly, so might have to disable this optimization in that case as well for now. Parquet reader was easier since most cases call `GetFirstFile()` in the binder, although this means that in the `main` branch, if the file pulled from `GetFirstFile()` gets pruned by hive filtering later, this might lead to unexpected results. Since this is existing behavior though I'm leaving that as is.

EDIT: I'm changing it so that instead of `GetAllFiles()` the CSV Schema Discovery will fetch the number of files to sniff based on options, which is still more than 1 or 2 (defaults to 10 I think), but is acceptable. I also encountered a CSV test that applies more generally, where a complex filter with OR and AND cannot be fully evaluated with a partial path, however it's possible to still be true if one side of the OR holds. This will take more work to do partial evaluation, so I think for now I'll leave it be. Once the globber reaches deep enough to get all the Hive values from the path it'll filter, which is still better than nothing.

https://github.com/duckdb/duckdb/blob/3de8d76ba83a82fb4337ca7cb8e4de2eab748561/test/sql/copy/csv/csv_hive.test#L92-L94

I've added a `hive_lazy_listing` parameter which currently defaults to true for this optimization, and sets to false when deactivated. Open to debate whether this should be on by default, but it feels more ergonomic to default to running the query as fast as correctly possible. I'm good either way though, as I can set it to true in `~/.duckdbrc` for myself, but would like to hear more thoughts on this.

Once I get existing tests passing, is it best to put up a PR for further discussion? Or should I hold off on that until ready for final review?

WIP: https://github.com/duckdb/duckdb/compare/main...xevix:hive-directory-pruning-2025-07-23?expand=1

**Comment by xevix** [comment]
Following up in #18518.
